### PR TITLE
Update benchmarks

### PR DESCRIPTION
**Pull request by miguelraz**
Hey all 👋🏾 

God hasn't killed me today so I've decided ~~to make it every slow code's problem~~ try and update these benchmarks.

The plan is to update the benchmarks now, and then to make an alternate implementation based on `AbstractKernels.jl` and compare it to a Futhark implementation, I think it will be interesting to compare lines of code, speed, and abstractional power (so that we may steal some good ones!)

Issues that I didn't know how to fix today:

* `"../../data/nn/list640k_64.txt": No such file or directory` on the `nn`
* I'm using `NVTX.@range` instead of `NVTX.@activate` but I'm willing to bet that's wrong, please correct me
* I don't know what `NVTX.stop()` did so I turned it off
* I get a bunch of `no method matching (CuArray{Float32})(::Int64))`

cc @vchuravy @maleadt

### CODE REVIEW

**Comment by maleadt** [comment]
> The plan is to update the benchmarks now, and then to make an alternate implementation based on `AbstractKernels.jl`

You mean KenrelAbstraction.jl, or are you creating a new kernel programming DSL?

> `"../../data/nn/list640k_64.txt": No such file or directory` on the `nn`

These should be generated by makefiles, e.g. https://github.com/JuliaParallel/rodinia/blob/a1debafca1d2145f287c2d8506fcf87ef6f1eb6e/data/nn/Makefile#L4-L5

> I'm using `NVTX.@range` instead of `NVTX.@activate` but I'm willing to bet that's wrong, please correct me
>
> I don't know what `NVTX.stop()` did so I turned it off

You can probably get rid of those. In addition, you can also get rid of the calls to `NVTX.@activate`, as it is lazily activated by NVTX.jl now. The only thing you should keep are the existing `NVTX.@range` annotations within `CUDA.@profile`.


> I get a bunch of `no method matching (CuArray{Float32})(::Int64))`

Those calls to e.g. `CuArray{Float32}(numRecords)` should become `CuArray{Float32}(undef, numRecords)`

**Comment by miguelraz** [comment]
Correct Tim, I meant `KernelAbstractions.jl`

OK, I
* Got rid of all `NVTX.stop()`
* slapped on the `undef`s where it made sense
* did `make` to get some data and had it all running silky smooth

I'm facing 2 glitches now:

1. `MAX_GRID_DIM_X` in the `nn.jl` file (where it used to be)
```julia
maxGridX = attribute(dev, CUDAnative.MAX_GRID_DIM_X)
```
and a few of the other attributes too.
2. `.buf` attribute not set in `streamcluster_cuda.jl`
```julia
    Mem.set!(work_mem_d.buf, UInt8(0), sizeof(work_mem_d)) # FIXME: needs wrapper
```

I think those are the only blockers for merging here. Pushing updated code now.

File: julia_cuda/nw/needle.jl [hunk 1]
```diff
@@ -17,37 +17,40 @@ const LIMIT = -999
 
 # static data
 const blosum62 = [
-     4 -1 -2 -2  0 -1 -1  0 -2 -1 -1 -1 -1 -2 -1  1  0 -3 -2  0 -2 -1  0 -4;
```
Comment (miguelraz) [inline julia_cuda/nw/needle.jl]: ugh, the formatter wrecked this.

Will punt until the last squash to revert.

**Comment by maleadt** [comment]
> 1. `MAX_GRID_DIM_X` in the `nn.jl` file (where it used to be)
> 
> 
> ```julia
> maxGridX = attribute(dev, CUDAnative.MAX_GRID_DIM_X)
> ```

That's `CUDA.DEVICE_ATTRIBUTE_MAX_GRID_DIM_X` now.

> 2. `.buf` attribute not set in `streamcluster_cuda.jl`
> 
> ```julia
>     Mem.set!(work_mem_d.buf, UInt8(0), sizeof(work_mem_d)) # FIXME: needs wrapper
> ```

Just use `fill!`; it should dispatch to `Mem.set!`: https://github.com/JuliaGPU/CUDA.jl/blob/d79adbfd090b0e51ccaf4c74710eaa610e0bf998/src/array.jl#L678-L692

**Comment by miguelraz** [comment]
@maleadt Modulo reformatting that big matrix that the formatter bonked, I think the work is finished for now.

Is there a robot to kick and `./run && ./verify` that all the benchmarks go according to plan?

**Comment by maleadt** [comment]
> Is there a robot to kick and `./run && ./verify` that all the benchmarks go according to plan?

Sadly not, this is from before the time we had GPU CI. I could have a look at adding some, will try to get to this some time this week.

**Comment by miguelraz** [comment]
Fair. What else should I do to get this PR merged then? Unbork the formatting on the matrix?

**Comment by maleadt** [comment]
> Unbork the formatting on the matrix?

Preferably. I don't mind the diff per se, but it just looked better before.

**Comment by miguelraz** [comment]
Done. Ready to merge then @maleadt

**Comment by maleadt** [comment]
I still don't like the formatting changes (it's very noisy and pretty opinionated to reformat an entire codebase as part of an unrelated PR), but I also don't care enough to have that hold up much needed improvements to the benchmarks. So thanks @miguelraz!

**Pull request merged**